ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java

```diff
@@ -517,22 +520,22 @@ private void reportInvalidBlock(final SignedBeaconBlock block, final BlockImport
 
   private void notifyForkChoiceUpdatedAndOptimisticSyncingChanged(
       final Optional<UInt64> proposingSlot) {
-    final ForkChoiceState forkChoiceState =
-        getForkChoiceStrategy()
-            .getForkChoiceState(
-                recentChainData.getCurrentEpoch().orElseThrow(),
-                recentChainData.getJustifiedCheckpoint().orElseThrow(),
-                recentChainData.getFinalizedCheckpoint().orElseThrow());
-
-    forkChoiceNotifier.onForkChoiceUpdated(forkChoiceState, proposingSlot);
-
-    if (optimisticSyncing
-        .map(oldValue -> !oldValue.equals(forkChoiceState.isHeadOptimistic()))
-        .orElse(true)) {
-      optimisticSyncing = Optional.of(forkChoiceState.isHeadOptimistic());
-      optimisticSyncSubscribers.deliver(
-          OptimisticHeadSubscriber::onOptimisticHeadChanged, forkChoiceState.isHeadOptimistic());
-    }
+    forkChoiceStateProvider
+        .getForkChoiceState()
+        .thenAccept(
```

Comment:
It doesn't seem good that this wound up becoming async. We could wind up not being on the fork choice thread or introducing other race conditions here. With the current code it will be sync but it would be way too easy to accidentally change that in the future and cause problems.

Hard to know exactly what to do about this, probably need to either have a `ForkChoiceStateProvider.getForkChoiceStateSync()` that just checks its already on the right thread and returns sync or make `ForkChoiceStateProvider` always expect to be on the right thread and make `ForkChoiceNotifier` handle getting onto the right thread.

Third option would be to make `ForkChoiceStateProvider` always sync and add a wrapper class that makes it async that `ForkChoiceNotifier` uses. They're all a bit distasteful really but I don't think we can safely assume a return `SafeFuture` completed synchronously.

Reply:
yeah I see. The executor should be returning with an immediately completed future if runs on the right thread.
But trusting the internals is not a good. I'm leaning to add a synchronous method...